crates/subspace-core-primitives/src/lib.rs

```diff
@@ -215,6 +222,144 @@ impl PosProof {
     pub const SIZE: usize = 17 * 8;
 }
 
+/// Proof of time key(input to the encryption).
+#[derive(Debug, Copy, Clone, Encode, Decode)]
+pub struct PotKey(PotBytes);
+
+impl PotKey {
+    /// Returns the key bytes.
+    #[inline]
+    pub fn bytes(&self) -> PotBytes {
+        self.0
+    }
+}
+
+impl From<PotBytes> for PotKey {
+    #[inline]
+    fn from(value: PotBytes) -> Self {
+        Self(value)
+    }
+}
+
+/// Proof of time seed (input to the encryption).
+#[derive(Debug, Copy, Clone, Encode, Decode)]
+pub struct PotSeed(PotBytes);
+
+impl PotSeed {
+    /// Returns the seed bytes.
+    #[inline]
+    pub fn bytes(&self) -> PotBytes {
+        self.0
+    }
+}
+
+impl From<PotBytes> for PotSeed {
+    #[inline]
+    fn from(value: PotBytes) -> Self {
+        Self(value)
+    }
+}
+
+impl From<BlockHash> for PotSeed {
+    #[inline]
+    fn from(block_hash: BlockHash) -> Self {
+        Self(pot_bytes_from_hash_256(block_hash))
+    }
+}
+
+impl From<(PotSeed, BlockHash)> for PotSeed {
+    #[inline]
+    fn from((seed, block_hash): (PotSeed, BlockHash)) -> Self {
+        let mut bytes = seed.bytes().to_vec();
+        bytes.append(&mut block_hash.to_vec());
+        let hash = blake2b_256_hash(&bytes);
+        Self(pot_bytes_from_hash_256(hash))
+    }
+}
+
+impl From<PotSeed> for PotKey {
+    fn from(seed: PotSeed) -> PotKey {
+        let seed_hash = blake2b_256_hash(&seed.bytes());
+        Self(pot_bytes_from_hash_256(seed_hash))
+    }
+}
+
+/// Proof of time ciphertext (output from the encryption).
+#[derive(Debug, Copy, Clone, Encode, Decode)]
+pub struct PotCipherText(PotBytes);
+
+impl PotCipherText {
+    /// Returns the encrypted bytes.
+    #[inline]
+    pub fn bytes(&self) -> PotBytes {
+        self.0
+    }
+}
+
+impl From<PotBytes> for PotCipherText {
+    #[inline]
+    fn from(value: PotBytes) -> Self {
+        Self(value)
+    }
+}
+
+/// Proof of time.
+/// TODO: versioning.
+#[derive(Debug, Clone, Encode, Decode)]
+pub struct ProofOfTime {
+    /// Slot the proof was evaluated for.
+    pub slot_number: SlotNumber,
+
+    /// The seed used for evaluation.
+    pub seed: PotSeed,
+
+    /// The encrypted outputs from each stage.
+    pub checkpoints: Vec<PotCipherText>,
+
+    /// Hash of last block at injection point.
+    pub injected_block_hash: BlockHash,
+}
+
+impl ProofOfTime {
+    /// Create the proof.
+    pub fn new(
+        slot_number: SlotNumber,
+        seed: PotSeed,
+        checkpoints: Vec<PotCipherText>,
+        injected_block_hash: BlockHash,
+    ) -> Self {
+        Self {
+            slot_number,
+            seed,
+            checkpoints,
+            injected_block_hash,
+        }
+    }
+
+    /// Returns the last check point.
+    pub fn output(&self) -> Option<PotCipherText> {
+        self.checkpoints.last().cloned()
+    }
+
+    /// Derives the global randomness from the output.
+    pub fn derive_global_randomness(&self) -> Option<Blake2b256Hash> {
```

Comment:
Alternatively you can create `Vec` wrapper that is guaranteed to contain at least one entry in it.